docs/sql-migration-guide.md

```diff
@@ -49,7 +49,7 @@ license: |
 
   - In Spark version 2.4 and earlier, float/double -0.0 is semantically equal to 0.0, but -0.0 and 0.0 are considered as different values when used in aggregate grouping keys, window partition keys and join keys. Since Spark 3.0, this bug is fixed. For example, `Seq(-0.0, 0.0).toDF("d").groupBy("d").count()` returns `[(0.0, 2)]` in Spark 3.0, and `[(0.0, 1), (-0.0, 1)]` in Spark 2.4 and earlier.
 
-  - In Spark version 2.4 and earlier, users can create a map with duplicated keys via built-in functions like `CreateMap`, `StringToMap`, etc. The behavior of map with duplicated keys is undefined, e.g. map look up respects the duplicated key appears first, `Dataset.collect` only keeps the duplicated key appears last, `MapKeys` returns duplicated keys, etc. Since Spark 3.0, these built-in functions will remove duplicated map keys with last wins policy. Users may still read map values with duplicated keys from data sources which do not enforce it (e.g. Parquet), the behavior will be undefined.
+  - In Spark version 2.4 and earlier, users can create a map with duplicated keys via built-in functions like `CreateMap`, `StringToMap`, etc. The behavior of map with duplicated keys is undefined, e.g. map look up respects the duplicated key appears first, `Dataset.collect` only keeps the duplicated key appears last, `MapKeys` returns duplicated keys, etc. Since Spark 3.0, new config `spark.sql.deduplicateMapKey.lastWinsPolicy.enabled` was added, with the default value `false`, Spark will throw RuntimeException while duplicated keys are found. If set to `true`, these built-in functions will remove duplicated map keys with last wins policy. Users may still read map values with duplicated keys from data sources which do not enforce it (e.g. Parquet), the behavior will be undefined.
```

Comment:
Hi, @cloud-fan , @gatorsmile , @rxin , @marmbrus .
So, `RuntimeException by default` is better in Apache Spark 3.0.0?

Reply:
The principle is still no failure by default, but version upgrade is a different case. We should try our best to avoid "silent result changing".

We may want to introduce a new category of configs, which is for migration-only, and should be removed in the next major version.

Thoughts? @srowen @viirya @maropu @bart-samwel

Reply:
Should we mention this `spark.sql.deduplicateMapKey.lastWinsPolicy.enabled` config is only useful for migration and could be removed in future version?

Reply:
This idea sounds reasonable. If we think a behavior change is correct but for migration reason, we don't want a silent behavior change, a config like this one is an option we can use to notify users such change.

Reply:
I agree with the idea to avoid  "silent result changing". Btw, we couldn't keep the old (spark 2.4) behaviour for duplicate keys by using a legacy option? If we couldn't do because of some reasons, the proposed one (the runtime exception) looks reasonable to me, too.

Reply:
The idea sounds fine.

Reply:
+1 for the idea of highlight this config is migration only.

Reply:
I'm okay with the rule, @cloud-fan . At this time, could you add a guideline into our website? Actually, many correctness bug fixes in the migration guide have been `silent result changing` with a legacy fallback config.

Reply:
Agree to have a guideline if we need the new rule, but `correctness bug fixes` might not the issue we discuss here? If it's a bug, why we need to have a legacy fallback config.

Reply:
What I have in mind is:
- if it's a bug fix (the previous result is definitely wrong), then no config is needed. If the impact is big, we can add a legacy config which is false by default.
- if it makes the behavior better, we should either add a config and use the old behavior by default, or fail by default and ask users to set config explicitly and pick the desired behavior.

I'm trying to think more cases, will send an email to the dev list soon.